# Wikipedia Notable Life Expectancies

# [Notebook 2 of 4: Data Cleaning](https://github.com/teresahanak/wikipedia-life-expectancy/blob/main/wp_life_expect_data_clean_thanak_2022_06_13.ipynb)

## Context

The


## Objective

The

### Data Dictionary

- Feature: Description

## Importing Necessary Libraries

In [1]:
# To structure code automatically
%load_ext nb_black

# To import/export sqlite databases
import sqlite3 as sql

# To help with reading and manipulating data
import pandas as pd
import numpy as np
import re

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)
# To define the maximum number of rows to be displayed in a dataframe
pd.set_option("display.max_rows", 200)

# To supress warnings
# import warnings

# warnings.filterwarnings("ignore")

# To set some visualization attributes
pd.set_option("max_colwidth", 150)

<IPython.core.display.Javascript object>

## Data Overview

### Reading, Sampling, and Checking Data Shape

In [2]:
# Reading the wp_life_expect_raw_complete dataset
conn = sql.connect("wp_life_expect_raw_complete.db")
data = pd.read_sql("SELECT * FROM wp_life_expect_raw_complete", conn)

# Making a working copy
df = data.copy()

# Checking the shape
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# Checking first 2 rows of the data
df.head(2)

There are 133900 rows and 6 columns.


month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21  
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12

<IPython.core.display.Javascript object>

In [3]:
# Checking last 2 rows of the data
df.tail(2)

month_year day           name  \
133898  June 2022   9  Oleg Moliboga   
133899  June 2022   9       Zou Jing   

                                                                         info  \
133898    , 69, Russian volleyball player, Olympic champion (1980) and coach.   
133899  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  
133898        https://en.wikipedia.org/wiki/Oleg_Moliboga              2  
133899  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3

<IPython.core.display.Javascript object>

In [4]:
# Checking a sample of the data
df.sample(5)

month_year day               name  \
107762  November 2019  11     Tam David-West   
17309    January 2001   2     George Ludford   
108976   January 2020   5  Sir Michael Stear   
107995  November 2019  22      Gugu Liberato   
110813     March 2020  18     Henri Richelet   

                                                             info  \
107762  , 83, Nigerian virologist, federal minister and academic.   
17309                      , 85, English professional footballer.   
108976           , 81, British Royal Air Force air chief marshal.   
107995                , 60, Brazilian television presenter, fall.   
110813                            , 75, French painter, COVID-19.   

                                                link num_references  
107762  https://en.wikipedia.org/wiki/Tam_David-West             12  
17309   https://en.wikipedia.org/wiki/George_Ludford              5  
108976   https://en.wikipedia.org/wiki/Michael_Stear             26  
107995   https://en.wikipedia.org/wiki/Gugu_Liberato             18  
110813  https://en.wikipedia.org/wiki/Henri_Richelet              4

<IPython.core.display.Javascript object>

#### Observations:
- There are 133,900 rows and 6 columns.
- `month_year` contains the month and year of death, while `day` contains the day of the month of death.
- `name` is the notable person's name.  It is a nominal feature that will not be used for analysis, but will be maintained for any referencing needs.
- `info` contains multiple items including the notable person's "age, country of citizenship at birth, subsequent country of citizenship (if applicable), reason for notability, (and) cause of death (if known)."
- `link` is the url to the notable person's individual Wikipedia page.  If such a page does not exist, there is either a non-working link (https://en.wikipedia.orgNone), or the link is to a page with a message that the page does not exist for that individual.  `link` is a unique identifier for all entries, except the 6 with the non-working link, which do have unique `name` values from each other.
- `num_references` contains the number of references on the notable person's individual Wikipedia page.  This feature serves as a proxy measure of notability.
- Prior to EDA, our task will be to extract the individual elements that are comined in `month_year` and `info` columns.

### Checking Data Types, Duplicates, and Null Values

In [5]:
# Checking data types and null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133894 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

In [6]:
# Checking duplicate rows
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [7]:
# Check percentage of null values by column
df.isnull().sum() / df.count() * 100

month_year        0.000000
day               0.000000
name              0.004481
info              0.000000
link              0.000000
num_references    0.000000
dtype: float64

<IPython.core.display.Javascript object>

In [8]:
# Checking number of missing values per row (not necessary here, but done to keep process standard)
df.isnull().sum(axis=1).value_counts()

0    133894
1         6
dtype: int64

<IPython.core.display.Javascript object>

#### Observations:
- Our dataset was saved to and read from the database without any hiccups.
- As expected, we have 6 entries that are missing `name`, but we will find it in their `info` values.
- All columns are currently of object type.  We will need to appropriately typecast them after separating the information in `month_year` and `info`.

## Data Cleaning

### Addressing Missing `name` Values

In [9]:
# Checking rows with missing name values
missing_name = df[df["name"].isna()]
missing_name

month_year day  name  \
18937      August 2001  11  None   
24985     January 2004  22  None   
27458       March 2005   1  None   
34077        July 2007  11  None   
64769   September 2013  29  None   
106613  September 2019  12  None   

                                                                                      info  \
18937   Kevin Kowalcyk, 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                Vincent Palmer, 37, British criminal.   
27458                                             Barry Stigler, 57, American voice actor.   
34077                                          Nana Gualdi, 75, German singer and actress.   
64769                                 Scott Workman, 47, American stuntman (, , ), cancer.   
106613                                                Thami Shobede, 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- These rows vary from the main set as there is a substring containing the person's name at the start of the `info` string.
- As there are so few rows missing `name`, let us address this issue first.

In [10]:
# For loop to copy name value from info value and remove name from info value
treat_rows = missing_name.index
for i in treat_rows:
    info = df.loc[i, "info"]
    info_lst = info.split(sep=",", maxsplit=1)

    name = info_lst[0].strip()
    df.loc[i, "name"] = name
    df.loc[i, "info"] = re.sub(name, "", info).strip()

# Re-check rows
df.loc[treat_rows, :]

month_year day            name  \
18937      August 2001  11  Kevin Kowalcyk   
24985     January 2004  22  Vincent Palmer   
27458       March 2005   1   Barry Stigler   
34077        July 2007  11     Nana Gualdi   
64769   September 2013  29   Scott Workman   
106613  September 2019  12   Thami Shobede   

                                                                        info  \
18937   , 2, known for eating a hamburger contaminated with E. coli O157:H7.   
24985                                                , 37, British criminal.   
27458                                            , 57, American voice actor.   
34077                                       , 75, German singer and actress.   
64769                                , 47, American stuntman (, , ), cancer.   
106613                                               , 31, Singer Songwriter   

                                link num_references  
18937   https://en.wikipedia.orgNone              0  
24985   https://en.wikipedia.orgNone              0  
27458   https://en.wikipedia.orgNone              0  
34077   https://en.wikipedia.orgNone              0  
64769   https://en.wikipedia.orgNone              0  
106613  https://en.wikipedia.orgNone              0

<IPython.core.display.Javascript object>

#### Observations:
- Missing `name` values have been addressed and those names have been removed from `info` values.

In [11]:
# Re-check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133900 entries, 0 to 133899
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   month_year      133900 non-null  object
 1   day             133900 non-null  object
 2   name            133900 non-null  object
 3   info            133900 non-null  object
 4   link            133900 non-null  object
 5   num_references  133900 non-null  object
dtypes: object(6)
memory usage: 6.1+ MB


<IPython.core.display.Javascript object>

#### Observations:
- We have no remaining missing values.
- Let us treat the `month_year` column next.

### Separating `month` and `year`

In [12]:
# Separating month and year into 2 columns and typecasting year as integer
df.loc[:, "year"] = df["month_year"].apply(lambda x: x.split(sep=" ")[1].strip())
df["year"] = df["year"].apply(lambda x: int(x))

df.loc[:, "month"] = df["month_year"].apply(lambda x: x.split(sep=" ")[0])
df.head(2)

month_year day              name  \
0  January 1994   1  William Chappell   
1  January 1994   1    Raymond Crotty   

                                                  info  \
0  , 86, British dancer, ballet designer and director.   
1         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

In [13]:
# Dropping month_year column
df.drop("month_year", axis=1, inplace=True)
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month  
0  1994  January  
1  1994  January

<IPython.core.display.Javascript object>

### Treating `info`

#### Checking a Sample

In [14]:
# Checking a sample of info
df.sample(5)

day          name  \
38487    3   Allan Deane   
103118  23    Rafi Eitan   
1598     8   Eijirō Tōno   
55171   16  Arthur Nadel   
124255  10  Frank Lobman   

                                                                                                                                      info  \
38487                                                                                                         , 85, New Zealand cricketer.   
103118  , 92, Israeli intelligence officer and politician, member of the Knesset (2006–2009) and Minister for Senior Citizens (2006–2009).   
1598                                                                                           , 86, Japanese actor (, , ), heart failure.   
55171                                                                                 , 80, American fund manager and convicted embezzler.   
124255                                                                                                   , 67, Surinamese-Dutch kickboxer.   

                                                       link num_references  \
38487             https://en.wikipedia.org/wiki/Allan_Deane              1   
103118             https://en.wikipedia.org/wiki/Rafi_Eitan             17   
1598    https://en.wikipedia.org/wiki/Eijir%C5%8D_T%C5%8Dno              2   
55171            https://en.wikipedia.org/wiki/Arthur_Nadel             44   
124255           https://en.wikipedia.org/wiki/Frank_Lobman              1   

        year      month  
38487   2008    October  
103118  2019      March  
1598    1994  September  
55171   2012      April  
124255  2021       June

<IPython.core.display.Javascript object>

#### Observations:
- First, let us check for any rows that are missing digits, and therefore the age target, within `info` and remove them.
- Also, it would be helpful to remove information contained within parentheses, as we will not be using it.

#### Checking and Dropping Rows Lacking Digits (and therefore Age Data) within `info`

In [15]:
# For loop to extract index of rows without digits in info value
remove_lst = []
for index in df.index:
    pattern = r"\d"
    if re.search(pattern, df.loc[index, "info"]) is None:
        remove_lst.append(index)
print(len(remove_lst), "rows")
df.loc[remove_lst, :].sample(5)

925 rows


day                           name  \
70848    9             Khayelihle Mathaba   
49408   16            Abdikadir Yusuf Aar   
133162   5             Justin Constantine   
14498    2  Muhammad Nasiruddin al-Albani   
119602  18                Joshua Kyeremeh   

                                                                                                            info  \
70848   , South African royal and politician, KwaZulu-Natal MLA, chieftain of the eMacambini, traffic collision.   
49408                                                                                , Somalian terrorist, shot.   
133162                                                                      , American military officer, cancer.   
14498                                                                                , Albanian Islamic scholar.   
119602                                                              , Ghanaian politician, kidney complications.   

                                                               link  \
70848              https://en.wikipedia.org/wiki/Khayelihle_Mathaba   
49408             https://en.wikipedia.org/wiki/Abdikadir_Yusuf_Aar   
133162             https://en.wikipedia.org/wiki/Justin_Constantine   
14498   https://en.wikipedia.org/wiki/Muhammad_Nasiruddin_al-Albani   
119602                https://en.wikipedia.org/wiki/Joshua_Kyeremeh   

       num_references  year    month  
70848               1  2014   August  
49408               2  2011    March  
133162             38  2022      May  
14498              37  1999  October  
119602              6  2021  January

<IPython.core.display.Javascript object>

In [16]:
# Dropping rows missing age data, resetting index, and checking new shape of df
df.drop(remove_lst, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132975, 7)

<IPython.core.display.Javascript object>

#### Observations:
- 925 rows were removed as they lacked the target data for `age`.

#### Removing Information within Parentheses in `info`

In [17]:
# Regular expression for parenthesis and its contents
pattern = r"\(.*\)"

# Subbing empty string for parentheses and stripping white space
df.loc[:, "info"] = df["info"].apply(lambda x: re.sub(pattern, "", x).strip())
df.loc[[21850, 125430], :]

day                 name  \
21850   11  Sir Michael Clapham   
125430  23       Jean-Luc Nancy   

                                                                                                      info  \
21850   , 90, British industrialist, president of the Confederation of British Industry from 1972 to 1974.   
125430                                                                           , 81, French philosopher.   

                                                                 link  \
21850   https://en.wikipedia.org/wiki/Michael_Clapham_(industrialist)   
125430                   https://en.wikipedia.org/wiki/Jean-Luc_Nancy   

       num_references  year     month  
21850              14  2002  November  
125430              8  2021    August

<IPython.core.display.Javascript object>

#### Observation:
- Parentheses and information within has been removed from `info`.
- Next, we will follow the Wikipedia-defined fields to divide the `info` values.

#### Splitting `info` on Commas into Separate Columns

In [18]:
# For loop to split info on commas and separate into respective new columns and removing leading/trailing white space and periods
for i, item in enumerate(df["info"]):
    info_lst = item.split(",")

    for j in range(len(info_lst)):
        df.loc[i, f"info_{j}"] = info_lst[j].strip(" .")

# Checking the first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_0 info_1           info_2                        info_3  \
0  1994  January            86   British dancer  ballet designer and director   
1  1994  January            68  Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [19]:
# Checking the last 2 rows
df.tail(2)

day           name  \
132973   9  Oleg Moliboga   
132974   9       Zou Jing   

                                                                         info  \
132973          , 69, Russian volleyball player, Olympic champion  and coach.   
132974  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132973        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132974  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_0 info_1                     info_2  \
132973  2022  June            69  Russian volleyball player   
132974  2022  June            86           Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132973                   Olympic champion  and coach    NaN    NaN    NaN   
132974  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11  
132973    NaN    NaN    NaN     NaN     NaN  
132974    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Observations:
- The `info` value is successfully divided and we can proceed through it column by column.
- We will check the set of values for the first two columns, for age.

### `info_0`

In [20]:
# Checking unique value counts
df["info_0"].value_counts()

                                                                       132946
Sir                                                                         6
92                                                                          2
Douglas Scott                                                               1
VC                                                                          1
Sir Woolwich West                                                           1
Sir Lord Justice of Appeal                                                  1
Dame MEP                                                                    1
83                                                                          1
Sir Governor-General                                                        1
Notable ice hockey players and coaches among the 44 killed in the :         1
Mike Alexander                                                              1
Colonel                                                         

<IPython.core.display.Javascript object>

#### Observations:
- The vast majority of rows have an empty string for this field.
- There is one row representing a group, rather than an individual, and we will drop it.
- We should verify the name and age information for the remainder of unique values in `info_0`.

#### Dropping Entry for Group

In [21]:
# Checking the entry representing a group
group_entry = df[
    df["info_0"]
    == "Notable ice hockey players and coaches among the 44 killed in the :"
]
group_entry

day                                  name  \
51130   7  2011 Lokomotiv Yaroslavl plane crash   

                                                                      info  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

                                                                     link  \
51130  https://en.wikipedia.org/wiki/2011_Lokomotiv_Yaroslavl_plane_crash   

      num_references  year      month  \
51130             95  2011  September   

                                                                    info_0  \
51130  Notable ice hockey players and coaches among the 44 killed in the :   

      info_1 info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
51130    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

      info_11  
51130     NaN

<IPython.core.display.Javascript object>

In [22]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132974, 19)

<IPython.core.display.Javascript object>

#### Examining Rows with Atypical `info_0` Values

In [23]:
# Examining rows with atypical info_0 values
list_to_check = df["info_0"].value_counts().index.to_list()

verify_df = pd.DataFrame()
for item in list_to_check[1:]:
    verify_df = pd.concat([verify_df, df[df["info_0"] == item]])
verify_df

day                     name  \
30349   25           Julian Bullard   
31140    1          Kyffin Williams   
34207   15             Jeremy Moore   
43606   29         Derek Hodgkinson   
67212    7             Richard Best   
67217    7             Thomas Hinde   
33086   28           David Turnbull   
117387  11         Gotthilf Fischer   
67517   21             Colin Turner   
67162    5               Robin Dunn   
67152    4    Elaine Kellett-Bowman   
59616   24      Kristján Jóhannsson   
52580   25           Moses Pitakaka   
42140    5                    Evile   
40355   22          Dudley Thornton   
34171    9       Sir Tasker Watkins   
33751   22             John Burnett   
30468    8           Peter Smithers   
14035    5                slapstick   
30363   26       Mahmoud al-Majzoub   
30358   26  Horondino José da Silva   
30210    6        Flight Lieutenant   
30188    2         Sam Mokuahi, Jr.   
29619   10             James Yancey   
18908   11              were killed   
18312   20            Demetreus Nix   
15873   15                      NRG   
67661   30            David Gibbons   

                                                                                                                 info  \
30349                                                                                     Sir , 78, British diplomat.   
31140                                                               Sir , 88, Welsh artist, lung and prostate cancer.   
34207                                    Sir , 79, British soldier, commander of UK land forces in the Falklands War.   
43606                                                                            Sir , 92, British air chief marshal.   
67212                                                             Sir , 80, British diplomat, Ambassador to Iceland .   
67217                                                                                     Sir , 88, British novelist.   
33086                                                                             . 92, American materials scientist.   
117387                                                                                  92, German choral conductor .   
67517                                                            Sir Woolwich West, 92, British politician, MP for  .   
67162                                                              Sir Lord Justice of Appeal, 96, British jurist,  .   
67152                                              Dame MEP, 90, British politician and barrister, MP for Lancaster .   
59616                                                                                  83, Icelandic Olympic athlete.   
52580                                                       Sir Governor-General, 66, Solomon Islander politician,  .   
42140                                                       Mike Alexander, 32, British bassist , pulmonary embolism.   
40355                                                                             Colonel , 89, British army officer.   
34171   VC, 88, British jurist and businessman, Lord Justice of Appeal and former WRU President, after short illness.   
33751                                                 Sir Edinburgh University, 85, British academic, Principal of  .   
30468                 Sir Winchester, 92, British politician, MP for Secretary General of the Council of Europe and .   
14035                                Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
30363                                  aka Abu Hamza, 41, Palestinian Islamic Jihad leader, assassination by bombing.   
30358                                     aka "Dino Sete Cordas", 88, Brazilian virtuouso of the seven-string guitar.   
30210                           Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
30188     aka "Sammy Steamboat", 71, American Hawaiian professional wrestler, complications 

<IPython.core.display.Javascript object>

#### Observations:
- The majority of rows contain additional aliases or titles within `info_0`, that we don't need, but we can leave in place for now.  
- There are a few rows that will need to be treated individually to correct the name value, as follows:
    1. Entry is for Mike Alexander whose band was Evile.
    2. Entry is for Herbert Wiere who performed slapstick.
    3. Entry is for Sarah-Jayne Mulvihill who was a Flight Lieutenant.
    4. Entry is for Douglass Scott who was killed by Demetreus Nix.
    5. Entry is for Kim Hwan-Sung who was a member of the band NRG.
- We can replace the `name` value with the `info_0` value for these rows as well as proceed with hard-coding the correct values for info_2 and info_3 fields to match the Wikipedia pattern, but staying true to the information scraped.
- The row with "Nearly 3" value for `info_0` represents a group, rather than an individual, so will be dropped, after treating the above rows.
- We can proceed to extract age from `info_0` for the few rows that contain it here instead of in `info_1`.

#### Treating 5 rows with Name in `info_0`

In [24]:
values_lst = [
    "Mike Alexander",
    "Herbert Wiere",
    "Sarah-Jayne Mulvihill",
    "Douglas Scott",
    "Kim Hwan-Sung",
]

<IPython.core.display.Javascript object>

In [25]:
# For loop to copy name from info_0 to name
for i in df[df["info_0"].isin(values_lst)].index.to_list():
    df.loc[i, "name"] = df.loc[i, "info_0"]

# Hard-coding info_2 and info_3 values for Kim Hwan-Sung
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/NRG_(South_Korean_band)"
].index.to_list()
df.loc[index, "info_2"] = "South Korean musician"

df.loc[index, "info_3"] = "respiratory illness"

# # Hard-coding info_2 and info_3 values for Douglass Scott
index = df[
    df["link"]
    == "https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1"
].index.to_list()
df.loc[index, "info_2"] = "student"

df.loc[index, "info_3"] = "murdered"

# # Hard-coding info_2 and info_3 values for Sarah-Jayne Mulvihill
index = df[
    df["link"] == "https://en.wikipedia.org/wiki/Flight_Lieutenant"
].index.to_list()
df.loc[index, "info_2"] = "British servicewoman"

df.loc[index, "info_3"] = "killed in action"

<IPython.core.display.Javascript object>

In [26]:
df[df["info_0"].isin(values_lst)]

day                   name  \
14035   5          Herbert Wiere   
15873  15          Kim Hwan-Sung   
18312  20          Douglas Scott   
30210   6  Sarah-Jayne Mulvihill   
42140   5         Mike Alexander   

                                                                                        info  \
14035       Herbert Wiere, 91, Austrian-born American Wiere Brothers comedian, member of the   
15873                                                       Kim Hwan-Sung, 19, A Member of .   
18312                                   Douglas Scott, 20, High-school student murdered by .   
30210  Sarah-Jayne Mulvihill, 32, first British servicewoman to be killed in action in Iraq.   
42140                              Mike Alexander, 32, British bassist , pulmonary embolism.   

                                                                                 link  \
14035                                         https://en.wikipedia.org/wiki/Slapstick   
15873                           https://en.wikipedia.org/wiki/NRG_(South_Korean_band)   
18312  https://en.wikipedia.org/w/index.php?title=Demetreus_Nix&action=edit&redlink=1   
30210                                 https://en.wikipedia.org/wiki/Flight_Lieutenant   
42140                                             https://en.wikipedia.org/wiki/Evile   

      num_references  year    month                 info_0 info_1  \
14035             16  1999   August          Herbert Wiere     91   
15873             27  2000     June          Kim Hwan-Sung     19   
18312              0  2001     June          Douglas Scott     20   
30210             12  2006      May  Sarah-Jayne Mulvihill     32   
42140             82  2009  October         Mike Alexander     32   

                                               info_2               info_3  \
14035  Austrian-born American Wiere Brothers comedian        member of the   
15873                           South Korean musician  respiratory illness   
18312                                         student             murdered   
30210                            British servicewoman     killed in action   
42140                                 British bassist   pulmonary embolism   

      info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11  
14035    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
15873    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
18312    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
30210    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  
42140    NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

#### Dropping Entry for Group

In [27]:
# Checking the entry representing a group
group_entry = df[df["info_0"] == "Nearly 3"]
group_entry

day         name  \
18908  11  were killed   

                                                               info  \
18908  Nearly 3,000 people September 11 attacks in the , including:   

                                                                       link  \
18908  https://en.wikipedia.org/wiki/Casualties_of_the_September_11_attacks   

      num_references  year      month    info_0  \
18908            176  2001  September  Nearly 3   

                                       info_1      info_2 info_3 info_4  \
18908  000 people September 11 attacks in the  including:    NaN    NaN   

      info_5 info_6 info_7 info_8 info_9 info_10 info_11  
18908    NaN    NaN    NaN    NaN    NaN     NaN     NaN

<IPython.core.display.Javascript object>

In [28]:
# Dropping group entry, resetting index, and checking new shape of df
df.drop(group_entry.index, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132973, 19)

<IPython.core.display.Javascript object>

#### Extracting `age` from `info_0`

In [29]:
# For loop to extract age from info_0
for i, item in enumerate(df["info_0"]):
    pattern = r"\d"
    if re.search(pattern, item):
        df.loc[i, "age"] = int(item)

# Checking one of these rows
df[df["info_0"] == "92"]

day              name                                 info  \
33085   28    David Turnbull  . 92, American materials scientist.   
117386  11  Gotthilf Fischer        92, German choral conductor .   

                                                                      link  \
33085   https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
117386                      https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year     month info_0                        info_1  \
33085               8  2007     April     92  American materials scientist   
117386              4  2020  December     92       German choral conductor   

       info_2 info_3 info_4 info_5 info_6 info_7 info_8 info_9 info_10  \
33085     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   
117386    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN     NaN   

       info_11   age  
33085      NaN  92.0  
117386     NaN  92.0

<IPython.core.display.Javascript object>

In [30]:
# Dropping info_0
df.drop("info_0", axis=1, inplace=True)

<IPython.core.display.Javascript object>

#### Observations:
- The new `age` column has been added successfully.
- We are finished processing `info_0`.

### `info_1`

#### Unique Values

In [31]:
# Checking unique values
df["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38',
       'American politician', '55/56', '87–88',
       '90 Swedish Olympic sprinte

<IPython.core.display.Javascript object>

#### Observations:
- There is a lot of variety in the format of the age data.
- Also, this field contains several values that we would expect in info_2 and beyond.
- Let us take the approach of extracting age values first, then examining rows with missing `age`.

#### Examining unique formats for age data

In [32]:
# Examining formats for age data
has_num = []
for i, item in enumerate(df["info_1"]):
    pattern = r"\d"
    if re.search(pattern, item) is not None:
        has_num.append(i)

df.loc[has_num, :]["info_1"].unique()

array(['86', '68', '87', '93', '79', '50', '88', '72', '81', '80', '90',
       '85', '92', '58', '54', '96', '49', '77', '76', '43', '35', '83',
       '31', '64', '57', '52', '84', '78', '70', '73', '67', '99', '33',
       '75', '66', '74', '62', '61', '82', '38', '47', '56', '91', '89',
       '94', '45', '65', '97', '63', '69', '37', '53', '46', '60', '26',
       '71', '25', '59', '39–40', '23', '95', '42', '32', '51', '41',
       '55', '44', '98', '39', '100', '27', '28', '40', '30', '48', '34',
       '29', '36', '111', '22', '104', '14', '21', '106', '105', '18',
       '101', '102', '20', '73-74', '84-85', '67-68', '85-86', '89-90',
       '103', '49-50', '87-88', '107', '48-49', '19', '71-72', '2',
       '82-83', '88-89', '11-12', '63-64', '64-65', '24', '17', '69-70',
       '52-53', '39-40', '80-81', '51-52', '70–71', '66–67', '59-60',
       '94-95', '15', '108', '76-77', '1995', '37-38', '55/56', '87–88',
       '90 Swedish Olympic sprinter', '6', '86-87', '62/63', '79

<IPython.core.display.Javascript object>

#### Observations:
- There are some specific rows that likely need to be dropped, with the following values for `info_1`:
    - 1995
    - 1997
    - German Olympic sailor [1]
    - Taiwanese failed assassin in the 3-19 shooting incident
    - 255
    - 176
    - the first wild bear in Germany in 170 years
    - c. 3500
    - common chimpanzee 55
    - Maltese 15
    - c.1000
    - Tree of the Year 150
- The dataset contains values for non-human species, so we will be on the lookout to drop those rows.
- The age formatting patterns include "age", "age1-age2" (short dash), "age1–age2" (long dash), 'age1/age2", "age ", 'age.', 'c. age', 'age/age 2nd digit', 'age or age', 'age?', 'early ages', 'age days', 'age-months', 'c.age', "age+", and "age months".
- We will need to address ages in days and months prior to those in years.

#### Examining Rows with Digits and Atypical Values for `info_2`

In [33]:
# List of atypical info_1 values for rows with digits
values_lst = [
    "1995",
    "1997",
    "German Olympic sailor [1]",
    "Taiwanese failed assassin in the 3-19 shooting incident",
    "255",
    "176",
    "the first wild bear in Germany in 170 years",
    "c. 3500",
    "common chimpanzee 55",
    "Maltese 15",
    "c.1000",
    "Tree of the Year 150",
]

df[df["info_1"].isin(values_lst)]

day                name  \
6046     9        Khaptad Baba   
6985     5  Wymberley D. Coerr   
10548   21       Pakoda Kadhar   
22722   17    Klaus Oldendorff   
25010   29      Chen Yi-hsiung   
29898   23             Adwaita   
30566   23             Harriet   
30591   26            Bear JJ1   
52996   16         The Senator   
55404    2              Oliver   
55455    5       Lucky Diamond   
86515    8  Pioneer Cabin Tree   
118228   5      Happy Man Tree   

                                                                                                                                       info  \
6046                                                                                                      , 1995, Nepalese spiritual saint.   
6985                                                                                              , 1995, American politician and diplomat.   
10548                                                                                                                 , 1997, Indian actor.   
22722                                                                                                           , German Olympic sailor [1]   
25010                                                                            , Taiwanese failed assassin in the 3-19 shooting incident.   
29898                          , 255 , tortoise claimant for world's oldest animal, reputedly a former pet of General Clive, liver failure.   
30566   , 176, Galápagos tortoise believed to be the third oldest animal in the world and allegedly owned by Charles Darwin, heart failure.   
30591                                                                         , the first wild bear in Germany in 170 years, shot to death.   
52996                                                                    , c. 3500, American pond cypress tree, largest in the world, fire.   
55404                                           , common chimpanzee 55, Congolese-born  noted for his upright stature and humanlike traits.   
55455                                                                   , Maltese 15, American Guinness World Record dog,  holder , cancer.   
86515                                                                                  , c.1000, American giant sequoia tree, storm damage.   
118228                                                                             , Tree of the Year 150, British plane tree,  , cut down.   

                                                     link num_references  \
6046           https://en.wikipedia.org/wiki/Khaptad_Baba              6   
6985     https://en.wikipedia.org/wiki/Wymberley_D._Coerr              3   
10548         https://en.wikipedia.org/wiki/Pakoda_Kadhar              1   
22722      https://en.wikipedia.org/wiki/Klaus_Oldendorff              1   
25010        https://en.wikipedia.org/wiki/Chen_Yi-hsiung              4   
29898               https://en.wikipedia.org/wiki/Adwaita              4   
30566    https://en.wikipedia.org/wiki/Harriet_(tortoise)             10   
30591              https://en.wikipedia.org/wiki/Bear_JJ1              1   
52996    https://en.wikipedia.org/wiki/The_Senator_(tree)             11   
55404   https://en.wikipedia.org/wiki/Oliver_(chimpanzee)             14   
55455   https://en.wikipedia.org/wiki/Lucky_Diamond_(dog)             13   
86515    https://en.wikipedia.org/wiki/Pioneer_Cabin_Tree             45   
118228       https://en.wikipedia.org/wiki/Happy_Man_Tree             23   

        year    month  \
6046    1996      May   
6985    1996  October   
10548   1998  January   
22722   2003    March   
25010   2004    March   
29898   2006    March   
30566   2006     June   
30591   2006     June   
52996   2012  January   
55404   2012     June   
55455   2012     June   
86515   2017  January   
118228  2021  January   

                                                         info_1  \
6046                                        

<IPython.core.display.Javascript object>

#### Observations:
- Age data is either missing or the entry is for a member of a non-human species.
- We will drop all of these rows.

#### Dropping Rows for Non-Human Entries or Entries Missing Age Data

In [34]:
# List of indexes to be dropped
drop_rows = df[df["info_1"].isin(values_lst)].index.to_list()

# Dropping rows, resetting index, and checking new shape of df
df.drop(drop_rows, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132960, 19)

<IPython.core.display.Javascript object>

### Extracting `age` from `info_1`

- Recall that the age formatting patterns include "age", "age1-age2" (short dash), "age1–age2" (long dash), 'age1/age2", "age ", 'age.', 'c. age', 'age/age 2nd digit', 'age or age', 'age?', 'early ages', 'age days', 'age-months', 'c.age', and "age months".
- We should address age in days and months first.

#### Step 1: Age in Years and Months

In [35]:
# Dictionary of patterns for days and months formats as keys and factor to convert to years
patterns = {
    r"(\d{1,3})( days)": 365,
    r"(\d{1,3})(-months)": 12,
    r"(\d{1,3})( months)": 12,
}

# For loop to check for age in days and months and convert to years and save in age.
# Also removes age in days/months from info_0.
for key, value in patterns.items():
    for i, item in enumerate(df["info_1"]):
        match = re.search(key, item)
        if match:
            age = int(match.group(1)) / value
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(key, "", df.loc[i, "info_1"])

# Checking updated rows
df[df["age"].notna()]

day                         name  \
28492   12  Susan Anne Catherine Torres   
30524   18          Chris and Cru Kahui   
33077   28               David Turnbull   
59604   24          Kristján Jóhannsson   
90456   28                 Charlie Gard   
117374  11             Gotthilf Fischer   

                                                                                                            info  \
28492   , 40 days, American baby born to Susan Torres, brain-dead woman, heart failure after intestinal surgery.   
30524                                                            , 3-months, New Zealand child homicide victims.   
33077                                                                        . 92, American materials scientist.   
59604                                                                             83, Icelandic Olympic athlete.   
90456                       , 11 months, British infant, subject of life support and parental rights case, MDDS.   
117374                                                                             92, German choral conductor .   

                                                                         link  \
28492                              https://en.wikipedia.org/wiki/Susan_Torres   
30524                       https://en.wikipedia.org/wiki/Chris_and_Cru_Kahui   
33077      https://en.wikipedia.org/wiki/David_Turnbull_(materials_scientist)   
59604   https://en.wikipedia.org/wiki/Kristj%C3%A1n_J%C3%B3hannsson_(athlete)   
90456                         https://en.wikipedia.org/wiki/Charlie_Gard_case   
117374                         https://en.wikipedia.org/wiki/Gotthilf_Fischer   

       num_references  year      month                        info_1  \
28492               3  2005  September                                 
30524              48  2006       June                                 
33077               8  2007      April  American materials scientist   
59604               2  2013    January     Icelandic Olympic athlete   
90456             125  2017       July                                 
117374              4  2020   December       German choral conductor   

                                    info_2  \
28492   American baby born to Susan Torres   
30524   New Zealand child homicide victims   
33077                                  NaN   
59604                                  NaN   
90456                       British infant   
117374                                 NaN   

                                                  info_3  \
28492                                   brain-dead woman   
30524                                                NaN   
33077                                                NaN   
59604                                                NaN   
90456   subject of life support and parental rights case   
117374                                               NaN   

                                        info_4 info_5 info_6 info_7 info_8  \
28492   heart failure after intestinal surgery    NaN    NaN    NaN    NaN   
30524                                      NaN    NaN    NaN    NaN    NaN   
33077                                      NaN    NaN    NaN    NaN    NaN   
59604                                      NaN    NaN    NaN    NaN    NaN   
90456                                     MDDS    NaN    NaN    NaN    NaN   
117374                                     NaN    NaN    NaN    NaN    NaN   

       info_9 info_10 info_11        age  
28492     NaN     NaN     NaN   0.109589  
30524     NaN     NaN     NaN   0.250000  
33077     NaN     NaN     NaN  92.000000  
59604     NaN     NaN     NaN  83.000000  
90456     NaN     NaN     NaN   0.916667  
117374    NaN     NaN     NaN  92.000000

<IPython.core.display.Javascript object>

#### Observations:
- We have successfully captured the age in days and months values and converted them to years.
- The other rows are in place already from our treatment of `info_0`.
- Next, we will extract the entries with straightforward single integer age values, including the formats: "age", "age ", 'age.'.
- Apparent estimates are excluded here to allow closer examination, as they are more likely to by atypical entries.

#### Step 2: Age as Single Integer (Excluding Estimates)

In [36]:
# List of patterns for age formats with single integer for age
patterns = [r"^(\d{1,3})$", r"^(\d{1,3})\s", r"^(\d{1,3})\.\s"]

# For loop to check age pattern in info_0, save age to age column, and remove from age from info_0
for i, item in enumerate(df["info_1"]):
    for pattern in patterns:
        match = re.search(pattern, item)
        if match:
            age = int(match.group(1))
            df.loc[i, "age"] = age
            df.loc[i, "info_1"] = re.sub(pattern, "", df.loc[i, "info_1"])

# Checking first 2 rows
df.head(2)

day              name                                                 info  \
0   1  William Chappell  , 86, British dancer, ballet designer and director.   
1   1    Raymond Crotty         , 68, Irish economist, writer, and academic.   

                                                      link num_references  \
0  https://en.wikipedia.org/wiki/William_Chappell_(dancer)             21   
1             https://en.wikipedia.org/wiki/Raymond_Crotty             12   

   year    month info_1           info_2                        info_3  \
0  1994  January          British dancer  ballet designer and director   
1  1994  January         Irish economist                        writer   

         info_4 info_5 info_6 info_7 info_8 info_9 info_10 info_11   age  
0           NaN    NaN    NaN    NaN    NaN    NaN     NaN     NaN  86.0  
1  and academic    NaN    NaN    NaN    NaN    NaN     NaN     NaN  68.0

<IPython.core.display.Javascript object>

In [37]:
# Checking last 2 rows
df.tail(2)

day           name  \
132958   9  Oleg Moliboga   
132959   9       Zou Jing   

                                                                         info  \
132958          , 69, Russian volleyball player, Olympic champion  and coach.   
132959  , 86, Chinese engineer, member of the Chinese Academy of Engineering.   

                                                     link num_references  \
132958        https://en.wikipedia.org/wiki/Oleg_Moliboga              2   
132959  https://en.wikipedia.org/wiki/Zou_Jing_(engineer)              3   

        year month info_1                     info_2  \
132958  2022  June         Russian volleyball player   
132959  2022  June                  Chinese engineer   

                                              info_3 info_4 info_5 info_6  \
132958                   Olympic champion  and coach    NaN    NaN    NaN   
132959  member of the Chinese Academy of Engineering    NaN    NaN    NaN   

       info_7 info_8 info_9 info_10 info_11   age  
132958    NaN    NaN    NaN     NaN     NaN  69.0  
132959    NaN    NaN    NaN     NaN     NaN  86.0

<IPython.core.display.Javascript object>

In [38]:
# Checking the number of remaining missing values for `age`
df["age"].isna().sum()

1014

<IPython.core.display.Javascript object>

#### Observations:
- The rows with single integer age data have been addressed.
- There are 1014 remaining missing values for `age`.
- Let us check the rows containing 'c.' in the age information.

#### Entries with Age Data Containing 'c.', '+', or '?' for Estimate

In [53]:
values = ["c.", "+", "?"]

# List of rows to check
rows_to_check = []

# For loop to add index of list containing values to list
for index in df[df["age"].isna()].index:
    if any(value in df.loc[index, "info_1"] for value in values):
        rows_to_check.append(index)

# Inspecting rows containing values
df.loc[rows_to_check, :]

day                                  name  \
12920   22                         Graciela Quan   
25360    4                   Tesfaye Gebre Kidan   
25440   18                          Paul Johnson   
25441   18                          Nek Mohammed   
26721   13                          Earl Cameron   
29736   28                            Peter Snow   
29832   15                              Humphrey   
30251   14                            Paul Marco   
31166    6          Mohammed Taha Mohammed Ahmed   
31448   11                       Benito Martínez   
32757   21                       Mohd. Ayub Khan   
33106    5     Prince Abdul-Majid bin Abdul-Aziz   
33977   21                             Čabulītis   
34609    6                  Sayed Mustafa Kazemi   
38001    1                        Arlene Sherman   
39737    2                         Halit Balamir   
40726    3                           Moloko Temo   
41381    5  Sheikha Hessa bint Salman Al Khalifa   
41382    5                      Baitullah Mehsud   
41838   14                Saleh Ali Saleh Nabhan   
43508   28              Mohammad-Reza Ali-Zamani   
43526   28                     Arash Rahmanipour   
44050    5                        Aminah Assilmi   
45134   24                       Abdolhamid Rigi   
47298   15                        Nimr al-Khatib   
48487   18                             Bob Tanna   
49151   12                         Buster Martin   
49410    1                            Anny Rüegg   
49903    8               Fazul Abdullah Mohammed   
53698   28                Abukar Hassan Mohamoud   
55671   24                       Lonesome George   
56256   31           Abdi Jeylani Malaq Marshale   
58486    9                              Ataa Oko   
61117   17                             Bi Kidude   
61259   23                         Mohammed Omar   
61984    4                           Luo Meizhen   
62418   27                      Muhammad Emin Er   
64614    3                                Reshma   
67361   22                    Hamza Abu al-Haija   
69694   22                          Glenn Jowitt   
70268   21                          Raed al Atar   
70948   26                                 Zelda   
73514    4      Sajida Mubarak Atrous al-Rishawi   
74316   19             Leandra Becerra Lumbreras   
74518   28                         Tuti Yusupova   
75369   13                           Bob Randall   
75598   27        Hassan Abdullah Hersi al-Turki   
78299    2                        Omar El-Hariri   
82150   21                        Akhtar Mansour   
94813   21                            Kakon Bibi   
99816   17                        Zura Karuhimbi   
105050  19                        Carlos Porrata   
109535  13                        Nasser Shabani   
128184  20                     Mamadú Iaia Djaló   

                                                                                                                           info  \
12920                                                                   , c. 79, Guatemalan lawyer and women's rights activist.   
25360                                                   , c. 69, Ethiopian general, defense minister and President of Ethiopia.   
25440                                                                       , c. 49, American hostage, decapitated by al-Qaeda.   
25441                 , c. 27, Pakistani tribal leader in Waziristan and key Taliban ally, killed by Pakistani military forces.   
26721                                                                                 , 89?, Canadian broadcaster and  anchor .   
29736                                                                , c. 70, New Zealand doctor who discovered "Tapanui flu" .   
29832                                                                    , c. 17, British Chief Mouser to the Cabinet Office, .   
30251                                                                                            ,

<IPython.core.display.Javascript object>

#### Observations:
- Most of the entries are for people, but there are also one or more entries for the following:
    - carp
    - racehorse
    - chimpanzee
    - flamingo
    - cat
    - turkey
- We will proceed to check `info_2` for these values and drop these and other rows representing members of these other species.

#### Checking for Cat, Racehorse, Chimpanzee, Carp, and Flamingo in `info_2`

In [54]:
# Defining pattern for re
species_pattern = r"\b(cat|racehorse|chimpanzee|carp|flamingo|turkey)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Checking number of rows containing pattern
print(f"There are {len(rows_to_check)} rows containing these values.")

There are 94 rows containing these values.


<IPython.core.display.Javascript object>

#### Observations:
- There are sufficient rows to warrant checking species by species.

#### Cat Entries per `info_2`

In [41]:
# Defining pattern for re
species_pattern = r"\b(cat)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                  name  \
18611    4         Lorenzo Music   
31856   29  Dewey Readmore Books   
38128   11              Scarlett   
39071    4                 India   
39689   20                 Socks   
41344   27                 Sybil   
47437   21          Prince Chunk   
54952    5                  Meow   
59825    4                Stewie   
63272    7     Buurtpoes Bledder   
66390   29          Colonel Meow   
70838   11       Hamish McHamish   
72424    4       Frank and Louie   
76284   22                  Tama   
90322   21                Stubbs   
92238    7               Paddles   
103399  14            Grumpy Cat   
107383   1               Lil Bub   
109614   3                    CC   

                                                                                                                              info  \
18611   , 64, American voice actor known for the voice of the cartoon cat Garfield, complications related to lung and bone cancer.   
31856                                                                                               , 19,  Library cat, euthanized   
38128                           , 13, American stray cat, name source of the Scarlett Award for Animal Heroism, animal euthanasia.   
39071                                                                                    , 18, American pet cat of George W. Bush.   
39689                                                           , 19, American Presidential cat of the Clinton family, euthanized.   
41344                                   , 3, British Downing Street cat, Chief Mouser to the Cabinet Office , after short illness.   
47437                                                                                     , 10, American obese cat, heart disease.   
54952                                                       , c. 2, American cat, heaviest cat at his time of death, lung failure.   
59825                                                                                 , 7–8, world's longest domestic cat, cancer.   
63272                                                                          , 1–2, domestic cat in Netherlands, motor accident.   
66390                                           , 2, American Himalayan-Persian cat, Guinness World Record holder for longest fur.   
70838                                                                               , 15, Scottish celebrity cat, chest infection.   
72424                                                  , 15, American-born Ragdoll cat, world's oldest diprosopus cat, euthanized.   
76284                                                    , 16, Japanese calico cat, stationmaster at Kishi Station, heart failure.   
90322                                                                    , 20, American cat, honorary mayor of Talkeetna, Alaska .   
92238                               , 1, New Zealand polydactyl cat, co-owned by Prime Minister Jacinda Ardern, traffic collision.   
103399                                           , 7, American internet celebrity cat, complications from urinary tract infection.   
107383                                                                                               , 8, American celebrity cat .   
109614                                                                 , 18, American tabby cat, first cloned pet, kidney failure.   

                                                      link num_references  \
18611          https://en.wikipedia.org/wiki/Lorenzo_Music              6   
31856   https://en.wikipedia.org/wiki/Dewey_Readmore_Books             26   
38128         https://en.wikipedia.org/wiki/Scarlett_(cat)              6   
39071            https://en.wikipedia.org/wiki/India_(cat)              7   
39689            https://en.wikipedia.org/wiki/Socks_(cat)             21   
41344            https://en.wikipedia.org/wiki/Sybil_(cat)              5   
47437           https://en.wikipedia.org/wiki/Prince_Chunk        

<IPython.core.display.Javascript object>

#### Observations:
- There is one person represented.  
- We can proceed to drop the others, using "cat", "cat of", and "cat in" patterns.

#### Dropping Entries for Cats per `info_2`

In [42]:
# List of re patterns to find
patterns = [r"\bcat$", r"\b(cat of|cat in)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
print(f"{len(rows_to_drop)} rows will be dropped.")

18 rows will be dropped.


<IPython.core.display.Javascript object>

In [43]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132942, 19)

<IPython.core.display.Javascript object>

####  Racehorse Entries per `info_2`

In [44]:
# Defining pattern for re
species_pattern = r"\bracehorse\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :].sample(10)

day               name  \
76382   29         Kauto Star   
120011  23         Sealy Hill   
17903   29   Dubai Millennium   
58986   28          Lord Avie   
56994    9             Monsun   
43494   22      Private Terms   
92364   15               Lure   
13689    1     Mr. Prospector   
28909    7             Fraise   
56875    1  Rainbows For Life   

                                                                                                                                              info  \
76382                                             , 15, French-born British-trained racehorse, dual winner of the Cheltenham Gold Cup, euthanized.   
120011                                                                                         , 17, Canadian Hall of Fame Thoroughbred racehorse.   
17903                                                                                                 , 5, Thoroughbred racehorse, grass sickness.   
58986                                                                  , 34, American thoroughbred racehorse, oldest living Eclipse Award winner .   
56994                                                                  , 22, German Thoroughbred racehorse, winner of the Aral-Pokal , euthanized.   
43494                                                                                           , 25, American Thoroughbred racehorse, euthanized.   
92364                                                                                                       , 28, American thoroughbred racehorse.   
13689                                                                             , 29, American thoroughbred racehorse, complications from colic.   
28909                                                                                                       , 17, American thoroughbred racehorse.   
56875   , 24, Canadian Thoroughbred racehorse, winner of the 1990 Summer Stakes, Cup and Saucer Stakes and Coronation Futurity Stakes, euthanized.   

                                                   link num_references  year  \
76382          https://en.wikipedia.org/wiki/Kauto_Star             48  2015   
120011         https://en.wikipedia.org/wiki/Sealy_Hill              3  2021   
17903    https://en.wikipedia.org/wiki/Dubai_Millennium             32  2001   
58986           https://en.wikipedia.org/wiki/Lord_Avie              7  2012   
56994              https://en.wikipedia.org/wiki/Monsun             18  2012   
43494       https://en.wikipedia.org/wiki/Private_Terms              2  2010   
92364        https://en.wikipedia.org/wiki/Lure_(horse)             22  2017   
13689      https://en.wikipedia.org/wiki/Mr._Prospector             31  1999   
28909              https://en.wikipedia.org/wiki/Fraise              9  2005   
56875   https://en.wikipedia.org/wiki/Rainbows_For_Life              0  2012   

            month info_1                                        info_2  \
76382        June                French-born British-trained racehorse   
120011   February         Canadian Hall of Fame Thoroughbred racehorse   
17903       April                               Thoroughbred racehorse   
58986    December                      American thoroughbred racehorse   
56994   September                        German Thoroughbred racehorse   
43494     January                      American Thoroughbred racehorse   
92364    November                      American thoroughbred racehorse   
13689        June                      American thoroughbred racehorse   
28909    November                      American thoroughbred racehorse   
56875   September                      Canadian Thoroughbred racehorse   

                                        info_3  \
76382   dual winner of the Cheltenham Gold Cup   
120011                                     NaN   
17903                           grass sickness   
58986       oldest living Eclipse Award winner   
56994                 winner of the Aral-Pokal

<IPython.core.display.Javascript object>

#### Observations:
- There are several entries for people involved in the racehorse business.
- Values that end in 'racehorse' and 'racehorse and sire' can be removed.

#### Dropping Entries for Racehorses per `info_2`

In [45]:
# List of re patterns to find
patterns = [r"\bracehorse$", r"\b(racehorse and sire)$"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
len(rows_to_drop)

349

<IPython.core.display.Javascript object>

In [46]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132593, 19)

<IPython.core.display.Javascript object>

####  Chimpanzee, Flamingo, and Carp Entries per `info_2`

In [47]:
# Defining pattern for re
species_pattern = r"\b(chimpanzee|flamingo|carp)\b"

# Empty list to collect indexes of rows containing pattern
rows_to_check = []

# For loop to add index of rows containing pattern to list
for index in df[df["info_2"].notna()].index:
    match = re.search(species_pattern, df.loc[index, "info_2"])
    if match:
        rows_to_check.append(index)

# Inpsecting rows with pattern
df.loc[rows_to_check, :]

day                 name  \
34557  30               Washoe   
39599  16               Travis   
41365   4               Benson   
45231   1  Heather the Leather   
66269  30              Greater   
76076  22     Don Featherstone   
92091  14          Little Mama   

                                                                                                                        info  \
34557  , c.42, American-trained African-born chimpanzee believed to be first non-human to acquire human language, influenza.   
39599                                                    , 13, American-born chimpanzee, television commercial animal, shot.   
41365                                                      , c. 25, British common carp, voted as Britain's Favourite Carp .   
45231                                                                                 , 50, British scaleless carp, old age.   
66269                                             , c. 83, Australian greater flamingo, world's oldest flamingo, euthanized.   
76076                                   , 79, American artist and inventor of the plastic pink flamingo, Lewy body dementia.   
92091                                                    , c.79, African-born chimpanzee , oldest on record, kidney failure.   

                                                          link num_references  \
34557        https://en.wikipedia.org/wiki/Washoe_(chimpanzee)             42   
39599        https://en.wikipedia.org/wiki/Travis_(chimpanzee)             69   
41365              https://en.wikipedia.org/wiki/Benson_(fish)              9   
45231        https://en.wikipedia.org/wiki/Heather_the_Leather              5   
66269         https://en.wikipedia.org/wiki/Greater_(flamingo)             12   
76076  https://en.wikipedia.org/wiki/Don_Featherstone_(artist)             10   
92091                https://en.wikipedia.org/wiki/Little_Mama              3   

       year     month info_1  \
34557  2007   October   c.42   
39599  2009  February          
41365  2009    August  c. 25   
45231  2010      June          
66269  2014   January  c. 83   
76076  2015      June          
92091  2017  November   c.79   

                                                                                                  info_2  \
34557  American-trained African-born chimpanzee believed to be first non-human to acquire human language   
39599                                                                           American-born chimpanzee   
41365                                                                                British common carp   
45231                                                                             British scaleless carp   
66269                                                                        Australian greater flamingo   
76076                                          American artist and inventor of the plastic pink flamingo   
92091                                                                            African-born chimpanzee   

                                  info_3          info_4 info_5 info_6 info_7  \
34557                          influenza             NaN    NaN    NaN    NaN   
39599       television commercial animal            shot    NaN    NaN    NaN   
41365  voted as Britain's Favourite Carp             NaN    NaN    NaN    NaN   
45231                            old age             NaN    NaN    NaN    NaN   
66269            world's oldest flamingo      euthanized    NaN    NaN    NaN   
76076                 Lewy body dementia             NaN    NaN    NaN    NaN   
92091                   oldest on record  kidney failure    NaN    NaN    NaN   

      info_8 info_9 info_10 info_11   age  
34557    NaN    NaN     NaN     NaN   NaN  
39599    NaN    NaN     NaN     NaN  13.0  
41365    NaN    NaN     NaN     NaN   NaN  
45231    NaN    NaN     NaN     NaN  50.0  
66269    NaN    NaN     NaN     NaN   NaN  
76076    NaN    NaN     

<IPython.core.display.Javascript object>

#### Observations:
- All of these rows are for animals, so we will remove them.

#### Dropping Entries for Chimpanzees, Carps, and Flamingos per `info_2`

In [48]:
# List of re patterns to find
patterns = [r"\b(chimpanzee|flamingo|carp)\b"]

# List to collect indexes of rows to drop
rows_to_drop = []

# For loop to find re pattern and add index of rows with pattern to list
for index in df[df["info_2"].notna()].index:
    for pattern in patterns:
        match = re.search(pattern, df.loc[index, "info_2"])
        if match:
            rows_to_drop.append(index)

# Checking number of rows added to list for dropping
len(rows_to_drop)

7

<IPython.core.display.Javascript object>

In [49]:
# Dropping rows, resetting index, and checking new shape of df
df.drop(rows_to_drop, inplace=True)
df.reset_index(inplace=True, drop=True)
df.shape

(132586, 19)

<IPython.core.display.Javascript object>

#### Observations:
- With those non-human entries addressed, we can return to processing `info_`.

In [50]:
df[df["age"].isna()]["info_1"].unique()

array(['39–40', '73-74', '84-85', '67-68', '85-86', '89-90', '49-50',
       '87-88', '48-49', '71-72', '82-83', '88-89', '11-12', '63-64',
       '64-65', '69-70', '52-53', '39-40', '80-81', '51-52', '70–71',
       '66–67', '59-60', '94-95', '76-77', '37-38', 'American politician',
       '55/56', '87–88', '86-87', '62/63', '79-80', '1996', '100/101',
       '97/98', '91/92', '66-67', '66/67', '38-39', '82–83', '88–89', '',
       '92–93', 'c. 79', '78/79', 'Prime Minister of Zaire', '85/86',
       '54/55', '50/51', '73/74', '67/68', '45–46', '34/35', '69–70',
       '75/76', 'King of Nepal', '89/90', '72/3',
       'member of the then National Assembly of Pakistan and Union Minister of Labor',
       '87/88', '28/29', '90/91', '52/53', '69/70', '73/4', '72-73',
       '52–53', '85–86', '82/83', 'Jules Engel', '94/95', '75-76', '76/7',
       '38-40', '60–61', 'c. 69', 'c. 49', 'c. 27', '43/4',
       'Vanuatuan president', '53/54', '55–56', '89?', '81-82', '54/5',
       'early 80s

<IPython.core.display.Javascript object>

In [ ]:
digits = [str(digit) for digit in np.arange(0, 10)]
rows_to_check = []
for index in df[df["age"].isna()].index:
    if any(digit in df.loc[index, "info_1"] for digit in digits):
        rows_to_check.append(index)
df.loc[rows_to_check, :].sample(10)

In [ ]:
df[df["info_1"] == "80. New Zealand Maori leader"]

In [51]:
df[df["info_1"] == "11+"]

day   name  \
70948  26  Zelda   

                                                                                            info  \
70948  , 11+, American wild turkey, resident of New York City's Battery Park, traffic collision.   

                                               link num_references  year  \
70948  https://en.wikipedia.org/wiki/Zelda_(turkey)             13  2014   

           month info_1                info_2  \
70948  September    11+  American wild turkey   

                                         info_3             info_4 info_5  \
70948  resident of New York City's Battery Park  traffic collision    NaN   

      info_6 info_7 info_8 info_9 info_10 info_11  age  
70948    NaN    NaN    NaN    NaN     NaN     NaN  NaN

<IPython.core.display.Javascript object>